bigtable/internal/testproxy/proxy.go

```diff
@@ -0,0 +1,963 @@
+	// creating the channel credentials.
+	if req.OverrideSslTargetName != "" {
+		d := grpc.WithAuthority(req.OverrideSslTargetName)
+		opts = append(opts, d)
+	}
+
+	// Case 1: No additional credentials provided
+	chc := req.GetChannelCredential()
+	if chc == nil {
+		return opts, nil
+	}
+	channelCreds, err := getChannelCredentials(chc, req.OverrideSslTargetName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Case 2: Only channel credentials provided
+	cc := req.CallCredential
+	if cc == nil {
+		d := grpc.WithTransportCredentials(channelCreds)
+		opts = append(opts, d)
+		return opts, nil
+	}
+
+	// Case 3: Both channel & call credentials provided
+	sa := cc.GetJsonServiceAccount()
+	clc, err := oauth.NewJWTAccessFromKey([]byte(sa))
+	if err != nil {
+		return nil, err
+	}
+
+	b := credentialsBundle{
+		channel: channelCreds,
+		call:    clc,
+	}
+
+	d := grpc.WithCredentialsBundle(b)
+	opts = append(opts, d)
+
+	return opts, nil
+}
+
+// getChannelCredentials extracts the channel credentials (credentials for use)
+// with all calls on this client.
+func getChannelCredentials(credsProto *pb.ChannelCredential, sslTargetName string) (credentials.TransportCredentials, error) {
+	var creds credentials.TransportCredentials
+	v := credsProto.GetValue()
+	switch t := v.(type) {
+	case *pb.ChannelCredential_Ssl:
+		pem := t.Ssl.GetPemRootCerts()
+
+		cert, err := x509.ParseCertificate([]byte(pem))
+		if err != nil {
+			return nil, err
+		}
+
+		pool := x509.NewCertPool()
+		pool.AddCert(cert)
+
+		creds = credentials.NewClientTLSFromCert(pool, sslTargetName)
+		if err != nil {
+			return nil, err
+		}
+	case *pb.ChannelCredential_None:
+		creds = insecure.NewCredentials()
+	default:
+		ctx := context.Background()
+		c, err := gauth.FindDefaultCredentials(ctx, []string{"https://www.googleapis.com/auth/cloud-platform"}...)
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO(developer): Determine how to pass this call option back to caller
+		option.WithTokenSource(c.TokenSource)
+
+		return nil, nil
+	}
+	return creds, nil
+}
+
+// goTestProxyServer represents an instance of the test proxy server. It keeps
+// a reference to individual clients instances (stored in a testClient object).
+type goTestProxyServer struct {
+	pb.UnimplementedCloudBigtableV2TestProxyServer
+	clientIDs map[string]testClient // clientIDs has all of the bigtable.Client objects under test
+}
+
+// CreateClient responds to the CreateClient RPC. This method adds a new client
+// instance to the goTestProxyServer
+func (s *goTestProxyServer) CreateClient(ctx context.Context, req *pb.CreateClientRequest) (*pb.CreateClientResponse, error) {
+	if req.ClientId == "" ||
+		req.DataTarget == "" ||
+		req.ProjectId == "" ||
+		req.InstanceId == "" {
+		return nil, stat.Error(codes.InvalidArgument,
+			fmt.Sprintf("%s must provide ClientId, DataTarget, ProjectId, and InstanceId", logLabel))
+	}
+
+	if _, exists := s.clientIDs[req.ClientId]; exists {
```

Comment:
Does this need to be thread safe?

Reply:
Done.

Reply:
Unless you're doing some time consuming IO in this method, you could just do this in all these methods:

``` golang
s.clientsLock.Lock()
defer s.clientsLock.Unlock()
```

Reply:
Oh, gotcha. Done.